## Data Overview

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Columns</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>acc_now_delinq</td>
      <td>The number of accounts on which the borrower is now delinquent.</td>
    </tr>
    <tr>
      <th>2</th>
      <td>acc_open_past_24mths</td>
      <td>Number of trades opened in the past 24 months.</td>
    </tr>
    <tr>
      <th>3</th>
      <td>addr_state</td>
      <td>The state provided by the borrower in the loan application.</td>
    </tr>
    <tr>
      <th>4</th>
      <td>all_util</td>
      <td>Balance to credit limit on all trades.</td>
    </tr>
    <tr>
      <th>5</th>
      <td>annual_inc</td>
      <td>The self-reported annual income provided by the borrower during registration.</td>
    </tr>
    <tr>
      <th>6</th>
      <td>annual_inc_joint</td>
      <td>The combined self-reported annual income provided by the co-borrowers during registration.</td>
    </tr>
    <tr>
      <th>7</th>
      <td>application_type</td>
      <td>Indicates whether the loan is an individual application or a joint application with two co-borrowers.</td>
    </tr>
    <tr>
      <th>8</th>
      <td>avg_cur_bal</td>
      <td>Average current balance of all accounts.</td>
    </tr>
    <tr>
      <th>9</th>
      <td>bc_open_to_buy</td>
      <td>Total open to buy on revolving bank cards.</td>
    </tr>
    <tr>
      <th>10</th>
      <td>bc_util</td>
      <td>Ratio of total current balance to high credit/credit limit for all bank card accounts.</td>
    </tr>
    <tr>
      <th>11</th>
      <td>chargeoff_within_12_mths</td>
      <td>Number of charge-offs within 12 months.</td>
    </tr>
    <tr>
      <th>12</th>
      <td>collection_recovery_fee</td>
      <td>Post charge-off collection fee.</td>
    </tr>
    <tr>
      <th>13</th>
      <td>collections_12_mths_ex_med</td>
      <td>Number of collections in 12 months excluding medical collections.</td>
    </tr>
    <tr>
      <th>14</th>
      <td>delinq_2yrs</td>
      <td>The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years.</td>
    </tr>
    <tr>
      <th>15</th>
      <td>delinq_amnt</td>
      <td>The past-due amount owed for the accounts on which the borrower is now delinquent.</td>
    </tr>
    <tr>
      <th>16</th>
      <td>desc</td>
      <td>Loan description provided by the borrower.</td>
    </tr>
    <tr>
      <th>17</th>
      <td>dti</td>
      <td>A ratio calculated using the borrower's total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower's self-reported monthly income.</td>
    </tr>
    <tr>
      <th>18</th>
      <td>dti_joint</td>
      <td>A ratio calculated using the co-borrowers' total monthly payments on the total debt obligations, excluding mortgages and the requested LC loan, divided by the co-borrowers' combined self-reported monthly income.</td>
    </tr>
    <tr>
      <th>19</th>
      <td>earliest_cr_line</td>
      <td>The month the borrower's earliest reported credit line was opened.</td>
    </tr>
    <tr>
      <th>20</th>
      <td>emp_length</td>
      <td>Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.</td>
    </tr>
    <tr>
      <th>21</th>
      <td>emp_title</td>
      <td>The job title supplied by the Borrower when applying for the loan.</td>
    </tr>
    <tr>
      <th>22</th>
      <td>fico_range_high</td>
      <td>The upper boundary range the borrower's FICO at loan origination belongs to.</td>
    </tr>
    <tr>
      <th>23</th>
      <td>fico_range_low</td>
      <td>The lower boundary range the borrower's FICO at loan origination belongs to.</td>
    </tr>
    <tr>
      <th>24</th>
      <td>funded_amnt</td>
      <td>The total amount committed to that loan at that point in time.</td>
    </tr>
    <tr>
      <th>25</th>
      <td>funded_amnt_inv</td>
      <td>The total amount committed by investors for that loan at that point in time.</td>
    </tr>
    <tr>
      <th>26</th>
      <td>grade</td>
      <td>LC assigned loan grade.</td>
    </tr>
    <tr>
      <th>27</th>
      <td>home_ownership</td>
      <td>The home ownership status provided by the borrower during registration. Our values are: RENT, OWN, MORTGAGE, OTHER.</td>
    </tr>
    <tr>
      <th>28</th>
      <td>id</td>
      <td>A unique LC assigned ID for the loan listing.</td>
    </tr>
    <tr>
      <th>29</th>
      <td>il_util</td>
      <td>Ratio of total current balance to high credit/credit limit on all installment accounts.</td>
    </tr>
    <tr>
      <th>30</th>
      <td>initial_list_status</td>
      <td>The initial listing status of the loan. Possible values are - W, F.</td>
    </tr>
    <tr>
      <th>31</th>
      <td>inq_fi</td>
      <td>Number of personal finance inquiries.</td>
    </tr>
    <tr>
      <th>32</th>
      <td>inq_last_12m</td>
      <td>Number of credit inquiries in the past 12 months.</td>
    </tr>
    <tr>
      <th>33</th>
      <td>inq_last_6mths</td>
      <td>The number of inquiries in the past 6 months (excluding auto and mortgage inquiries).</td>
    </tr>
    <tr>
      <th>34</th>
      <td>installment</td>
      <td>The monthly payment owed by the borrower if the loan originates.</td>
    </tr>
    <tr>
      <th>35</th>
      <td>int_rate</td>
      <td>Interest Rate on the loan.</td>
    </tr>
    <tr>
      <th>36</th>
      <td>issue_d</td>
      <td>The month which the loan was funded.</td>
    </tr>
    <tr>
      <th>37</th>
      <td>last_credit_pull_d</td>
      <td>The most recent month LC pulled credit for this loan.</td>
    </tr>
    <tr>
      <th>38</th>
      <td>last_fico_range_high</td>
      <td>The upper boundary range the borrower's last FICO pulled belongs to.</td>
    </tr>
    <tr>
      <th>39</th>
      <td>last_fico_range_low</td>
      <td>The lower boundary range the borrower's last FICO pulled belongs to.</td>
    </tr>
    <tr>
      <th>40</th>
      <td>last_pymnt_amnt</td>
      <td>Last total payment amount received.</td>
    </tr>
    <tr>
      <th>41</th>
      <td>last_pymnt_d</td>
      <td>Last month payment was received.</td>
    </tr>
    <tr>
      <th>42</th>
      <td>loan_amnt</td>
      <td>The listed amount of the loan applied for by the borrower. If at some point in time, the credit department reduces the loan amount, then it will be reflected in this value.</td>
    </tr>
    <tr>
      <th>43</th>
      <td>loan_status</td>
      <td>Current status of the loan.</td>
    </tr>
    <tr>
      <th>44</th>
      <td>max_bal_bc</td>
      <td>Maximum current balance owed on all revolving accounts.</td>
    </tr>
    <tr>
      <th>45</th>
      <td>member_id</td>
      <td>A unique LC assigned ID for the borrower member.</td>
    </tr>
    <tr>
      <th>46</th>
      <td>mo_sin_old_il_acct</td>
      <td>Months since oldest bank installment account opened.</td>
    </tr>
    <tr>
      <th>47</th>
      <td>mo_sin_old_rev_tl_op</td>
      <td>Months since oldest revolving account opened.</td>
    </tr>
    <tr>
      <th>48</th>
      <td>mo_sin_rcnt_rev_tl_op</td>
      <td>Months since most recent revolving account opened.</td>
    </tr>
    <tr>
      <th>49</th>
      <td>mo_sin_rcnt_tl</td>
      <td>Months since most recent account opened.</td>
    </tr>
    <tr>
      <th>50</th>
      <td>mort_acc</td>
      <td>Number of mortgage accounts.</td>
    </tr>
    <tr>
      <th>51</th>
      <td>mths_since_last_delinq</td>
      <td>The number of months since the borrower's last delinquency.</td>
    </tr>
    <tr>
      <th>52</th>
      <td>mths_since_last_major_derog</td>
      <td>Months since most recent 90-day or worse rating.</td>
    </tr>
    <tr>
      <th>53</th>
      <td>mths_since_last_record</td>
      <td>The number of months since the last public record.</td>
    </tr>
    <tr>
      <th>54</th>
      <td>mths_since_rcnt_il</td>
      <td>Months since most recent installment accounts opened.</td>
    </tr>
    <tr>
      <th>55</th>
      <td>mths_since_recent_bc</td>
      <td>Months since most recent bank card account opened.</td>
    </tr>
    <tr>
      <th>56</th>
      <td>mths_since_recent_bc_dlq</td>
      <td>Months since most recent bank card delinquency.</td>
    </tr>
    <tr>
      <th>57</th>
      <td>mths_since_recent_inq</td>
      <td>Months since most recent inquiry.</td>
    </tr>
    <tr>
      <th>58</th>
      <td>mths_since_recent_revol_delinq</td>
      <td>Months since most recent revolving delinquency.</td>
    </tr>
    <tr>
      <th>59</th>
      <td>next_pymnt_d</td>
      <td>Next scheduled payment date.</td>
    </tr>
    <tr>
      <th>60</th>
      <td>num_accts_ever_120_pd</td>
      <td>Number of accounts ever 120 or more days past due.</td>
    </tr>
    <tr>
      <th>61</th>
      <td>num_actv_bc_tl</td>
      <td>Number of currently active bank card accounts.</td>
    </tr>
    <tr>
      <th>62</th>
      <td>num_actv_rev_tl</td>
      <td>Number of currently active revolving trades.</td>
    </tr>
    <tr>
      <th>63</th>
      <td>num_bc_sats</td>
      <td>Number of satisfactory bank card accounts.</td>
    </tr>
    <tr>
      <th>64</th>
      <td>num_bc_tl</td>
      <td>Number of bank card accounts.</td>
    </tr>
    <tr>
      <th>65</th>
      <td>num_il_tl</td>
      <td>Number of installment accounts.</td>
    </tr>
    <tr>
      <th>66</th>
      <td>num_op_rev_tl</td>
      <td>Number of open revolving accounts.</td>
    </tr>
    <tr>
      <th>67</th>
      <td>num_rev_accts</td>
      <td>Number of revolving accounts.</td>
    </tr>
    <tr>
      <th>68</th>
      <td>num_rev_tl_bal_gt_0</td>
      <td>Number of revolving trades with a balance >0.</td>
    </tr>
    <tr>
      <th>69</th>
      <td>num_sats</td>
      <td>Number of satisfactory accounts.</td>
    </tr>
    <tr>
      <th>70</th>
      <td>num_tl_120dpd_2m</td>
      <td>Number of accounts currently 120 days past due (updated in the past 2 months).</td>
    </tr>
    <tr>
      <th>71</th>
      <td>num_tl_30dpd</td>
      <td>Number of accounts currently 30 days past due (updated in the past 2 months).</td>
    </tr>
    <tr>
      <th>72</th>
      <td>num_tl_90g_dpd_24m</td>
      <td>Number of accounts 90 or more days past due in the last 24 months.</td>
    </tr>
    <tr>
      <th>73</th>
      <td>num_tl_op_past_12m</td>
      <td>Number of accounts opened in the past 12 months.</td>
    </tr>
    <tr>
      <th>74</th>
      <td>open_acc</td>
      <td>The number of open credit lines in the borrower's credit file.</td>
    </tr>
    <tr>
      <th>75</th>
      <td>open_acc_6m</td>
      <td>Number of open trades in the last 6 months.</td>
    </tr>
    <tr>
      <th>76</th>
      <td>open_il_12m</td>
      <td>Number of installment accounts opened in the past 12 months.</td>
    </tr>
    <tr>
      <th>77</th>
      <td>open_il_24m</td>
      <td>Number of installment accounts opened in the past 24 months.</td>
    </tr>
    <tr>
      <th>78</th>
      <td>open_il_6m</td>
      <td>Number of currently active installment trades.</td>
    </tr>
    <tr>
      <th>79</th>
      <td>open_rv_12m</td>
      <td>Number of revolving trades opened in the past 12 months.</td>
    </tr>
    <tr>
      <th>80</th>
      <td>open_rv_24m</td>
      <td>Number of revolving trades opened in the past 24 months.</td>
    </tr>
    <tr>
      <th>81</th>
      <td>out_prncp</td>
      <td>Remaining outstanding principal for the total amount funded.</td>
    </tr>
    <tr>
      <th>82</th>
      <td>out_prncp_inv</td>
      <td>Remaining outstanding principal for the portion of the total amount funded by investors.</td>
    </tr>
    <tr>
      <th>83</th>
      <td>pct_tl_nvr_dlq</td>
      <td>Percent of trades never delinquent.</td>
    </tr>
    <tr>
      <th>84</th>
      <td>percent_bc_gt_75</td>
      <td>Percentage of all bank card accounts > 75% of the limit.</td>
    </tr>
    <tr>
      <th>85</th>
      <td>policy_code</td>
      <td>"Publicly available policy_code=1
        New products not publicly available policy_code=2"</td>
    </tr>
    <tr>
      <th>86</th>
      <td>pub_rec</td>
      <td>Number of derogatory public records.</td>
    </tr>
    <tr>
      <th>87</th>
      <td>pub_rec_bankruptcies</td>
      <td>Number of public record bankruptcies.</td>
    </tr>
    <tr>
      <th>88</th>
      <td>purpose</td>
      <td>A category provided by the borrower for the loan request.</td>
    </tr>
    <tr>
      <th>89</th>
      <td>pymnt_plan</td>
      <td>Indicates if a payment plan has been put in place for the loan.</td>
    </tr>
    <tr>
      <th>90</th>
      <td>recoveries</td>
      <td>Post charge-off gross recovery.</td>
    </tr>
    <tr>
      <th>91</th>
      <td>revol_bal</td>
      <td>Total credit revolving balance.</td>
    </tr>
    <tr>
      <th>92</th>
      <td>revol_util</td>
      <td>Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit.</td>
    </tr>
    <tr>
      <th>93</th>
      <td>sub_grade</td>
      <td>LC assigned loan subgrade.</td>
    </tr>
    <tr>
      <th>94</th>
      <td>tax_liens</td>
      <td>Number of tax liens.</td>
    </tr>
    <tr>
      <th>95</th>
      <td>term</td>
      <td>The number of payments on the loan. Values are in months and can be either 36 or 60.</td>
    </tr>
    <tr>
      <th>96</th>
      <td>title</td>
      <td>The loan title provided by the borrower.</td>
    </tr>
    <tr>
      <th>97</th>
      <td>tot_coll_amt</td>
      <td>Total collection amounts ever owed.</td>
    </tr>
    <tr>
      <th>98</th>
      <td>tot_cur_bal</td>
      <td>Total current balance of all accounts.</td>
    </tr>
    <tr>
      <th>99</th>
      <td>tot_hi_cred_lim</td>
      <td>Total high credit/credit limit.</td>
    </tr>
    <tr>
      <th>100</th>
      <td>total_acc</td>
      <td>The total number of credit lines currently in the borrower's credit file.</td>
    </tr>
    <tr>
      <th>101</th>
      <td>total_bal_ex_mort</td>
      <td>Total credit balance excluding mortgage.</td>
    </tr>
    <tr>
      <th>102</th>
      <td>total_bc_limit</td>
      <td>Total bank card high credit/credit limit.</td>
    </tr>
    <tr>
      <th>103</th>
      <td>total_cu_tl</td>
      <td>Total current unused revolving credit lines.</td>
    </tr>
    <tr>
      <th>104</th>
      <td>total_il_high_credit_limit</td>
      <td>Total installment high credit/credit limit.</td>
    </tr>
    <tr>
      <th>105</th>
      <td>total_pymnt</td>
      <td>Payments received to date for total amount funded.</td>
    </tr>
    <tr>
      <th>106</th>
      <td>total_pymnt_inv</td>
      <td>Payments received to date for the portion of the total amount funded by investors.</td>
    </tr>
    <tr>
      <th>107</th>
      <td>total_rec_int</td>
      <td>Interest received to date.</td>
    </tr>
    <tr>
      <th>108</th>
      <td>total_rec_late_fee</td>
      <td>Late fees received to date.</td>
    </tr>
    <tr>
      <th>109</th>
      <td>total_rec_prncp</td>
      <td>Principal received to date.</td>
    </tr>
    <tr>
      <th>110</th>
      <td>total_rev_hi_lim</td>
      <td>Total revolving high credit/credit limit.</td>
    </tr>
    <tr>
      <th>111</th>
      <td>url</td>
      <td>URL for the LC page with listing data.</td>
    </tr>
    <tr>
      <th>112</th>
      <td>verified_status_joint</td>
      <td>Indicates if the co-borrowers' joint income was verified by LC.</td>
    </tr>
    <tr>
      <th>113</th>
      <td>zip_code</td>
      <td>The first 3 numbers of the zip code provided by the borrower in the loan application.</td>
    </tr>
  </tbody>
</table>

## Loading and Setup

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
# Filter warnings
from warnings import filterwarnings
filterwarnings('ignore')

/Users/rachelliao/anaconda3/envs/ensemble/lib/python3.8/site-packages/xgboost/compat.py:93: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import MultiIndex, Int64Index


In [2]:
# Load the dataset
df = pd.read_parquet('dataset/LoanData_2007_to_2018.parquet')

Set up feature information for easy lookup throughout the notebook:

In [3]:
# Load the data dictionary
data_info = pd.read_csv('dataset/data_dictionary.csv', index_col='LoanStatNew')

In [4]:
# Testing out the feature information lookup
print(data_info.loc['annual_inc']['Description'])

The self-reported annual income provided by the borrower during registration.


In [5]:
# Set up self-defined function for feature information
def col_info(col_name):
    print(data_info.loc[col_name]['Description'])

In [6]:
# Testing out the self-defined function
col_info('annual_inc')

The self-reported annual income provided by the borrower during registration.


## Assessment

In [7]:
# Set display options to show all columns and rows
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [8]:
# Print the shape of the dataset
df.shape

(22511, 151)

In [9]:
# Print the first 5 rows of the dataset
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  68407277        NaN     3600.0       3600.0           3600.0   36 months   
1  68355089        NaN    24700.0      24700.0          24700.0   36 months   
2  68341763        NaN    20000.0      20000.0          20000.0   60 months   
3  66310712        NaN    35000.0      35000.0          35000.0   60 months   
4  68476807        NaN    10400.0      10400.0          10400.0   60 months   

   int_rate  installment grade sub_grade                    emp_title  \
0     13.99       123.03     C        C4                      leadman   
1     11.99       820.28     C        C1                     Engineer   
2     10.78       432.66     B        B4                 truck driver   
3     14.85       829.90     C        C5  Information Systems Officer   
4     22.45       289.91     F        F1          Contract Specialist   

  emp_length home_ownership  annual_inc verification_status   issue_d  \
0  10+ years       MORTGAGE     55000.0        Not Verified  Dec-2015   
1  10+ years       MORTGAGE     65000.0        Not Verified  Dec-2015   
2  10+ years       MORTGAGE     63000.0        Not Verified  Dec-2015   
3  10+ years       MORTGAGE    110000.0     Source Verified  Dec-2015   
4    3 years       MORTGAGE    104433.0     Source Verified  Dec-2015   

  loan_status pymnt_plan                                                url  \
0  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
1  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
2  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
3     Current          n  https://lendingclub.com/browse/loanDetail.acti...   
4  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   

   desc             purpose               title zip_code addr_state    dti  \
0  None  debt_consolidation  Debt consolidation    190xx         PA   5.91   
1  None      small_business            Business    577xx         SD  16.06   
2  None    home_improvement                None    605xx         IL  10.78   
3  None  debt_consolidation  Debt consolidation    076xx         NJ  17.06   
4  None      major_purchase      Major purchase    174xx         PA  25.37   

   delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
0          0.0         Aug-2003           675.0            679.0   
1          1.0         Dec-1999           715.0            719.0   
2          0.0         Aug-2000           695.0            699.0   
3          0.0         Sep-2008           785.0            789.0   
4          1.0         Jun-1998           695.0            699.0   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0             1.0                    30.0                     NaN       7.0   
1             4.0                     6.0                     NaN      22.0   
2             0.0                     NaN                     NaN       6.0   
3             0.0                     NaN                     NaN      13.0   
4             3.0                    12.0                     NaN      12.0   

   pub_rec  revol_bal  revol_util  total_acc initial_list_status  out_prncp  \
0      0.0     2765.0        29.7       13.0                   w       0.00   
1      0.0    21470.0        19.2       38.0                   w       0.00   
2      0.0     7869.0        56.2       18.0                   w       0.00   
3      0.0     7802.0        11.6       17.0                   w   15897.65   
4      0.0    21929.0        64.5       35.0                   w       0.00   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   4421.723917          4421.72          3600.00   
1           0.00  25679.660000         25679.66         24700.00   
2           0.00  22705.924294         22705.92         20000.00   
3       15897.65  31464.010000         31464.01         19102.35   
4           0

In [10]:
# Display the info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22511 entries, 0 to 22510
Columns: 151 entries, id to settlement_term
dtypes: float64(114), int64(1), object(36)
memory usage: 25.9+ MB


In [11]:
# Display the data types of the dataset
df.dtypes

id                                              int64
member_id                                     float64
loan_amnt                                     float64
funded_amnt                                   float64
funded_amnt_inv                               float64
term                                           object
int_rate                                      float64
installment                                   float64
grade                                          object
sub_grade                                      object
emp_title                                      object
emp_length                                     object
home_ownership                                 object
annual_inc                                    float64
verification_status                            object
issue_d                                        object
loan_status                                    object
pymnt_plan                                     object
url                         

In [12]:
# Check the number of unique values in each column
df.nunique()

id                                            22511
member_id                                         0
loan_amnt                                      1140
funded_amnt                                    1140
funded_amnt_inv                                1190
term                                              2
int_rate                                         64
installment                                    6915
grade                                             7
sub_grade                                        35
emp_title                                     10992
emp_length                                       11
home_ownership                                    3
annual_inc                                     2442
verification_status                               3
issue_d                                           1
loan_status                                       6
pymnt_plan                                        2
url                                           22511
desc        

In [13]:
# Get a statistical summary of the dataset
df.describe()

id  member_id     loan_amnt   funded_amnt  funded_amnt_inv  \
count  2.251100e+04        0.0  22511.000000  22511.000000     22511.000000   
mean   6.756447e+07        NaN  15031.787793  15031.787793     15025.527520   
std    1.843754e+06        NaN   8734.651784   8734.651784      8729.680902   
min    3.617740e+05        NaN   1000.000000   1000.000000      1000.000000   
25%    6.729818e+07        NaN   8000.000000   8000.000000      8000.000000   
50%    6.752576e+07        NaN  13975.000000  13975.000000     13900.000000   
75%    6.811268e+07        NaN  20000.000000  20000.000000     20000.000000   
max    6.861706e+07        NaN  35000.000000  35000.000000     35000.000000   

           int_rate   installment    annual_inc           dti   delinq_2yrs  \
count  22511.000000  22511.000000  2.251100e+04  22509.000000  22510.000000   
mean      12.255020    436.321286  7.820695e+04     19.423974      0.335584   
std        4.213027    249.411778  6.125830e+04     11.006613      0.887691   
min        5.320000     30.540000  0.000000e+00      0.000000      0.000000   
25%        9.170000    255.040000  4.800000e+04     12.700000      0.000000   
50%       11.990000    379.390000  6.700000e+04     18.840000      0.000000   
75%       14.480000    579.130000  9.500000e+04     25.660000      0.000000   
max       28.990000   1354.660000  3.964280e+06    999.000000     15.000000   

       fico_range_low  fico_range_high  inq_last_6mths  \
count    22510.000000     22510.000000    22510.000000   
mean       695.075522       699.075611        0.608796   
std         31.228687        31.229115        0.887993   
min        660.000000       664.000000        0.000000   
25%        670.000000       674.000000        0.000000   
50%        690.000000       694.000000        0.000000   
75%        710.000000       714.000000        1.000000   
max        845.000000       850.000000        5.000000   

       mths_since_last_delinq  mths_since_last_record      open_acc  \
count            11637.000000              4038.00000  22510.000000   
mean                34.162929                65.69787     12.140116   
std                 21.863500                23.81295      5.776034   
min                  0.000000                 0.00000      1.000000   
25%                 15.000000                51.00000      8.000000   
50%                 31.000000                68.00000     11.000000   
75%                 49.000000                82.00000     15.000000   
max                120.000000               120.00000     65.000000   

            pub_rec      revol_bal   revol_util     total_acc     out_prncp  \
count  22510.000000   22510.000000  22499.00000  22510.000000  22510.000000   
mean       0.238250   17388.656908     51.57675     25.538783    994.961977   
std        0.646684   22622.637023     24.22553     12.195480   3018.661632   
min        0.000000       0.000000      0.00000      4.000000      0.000000   
25%        0.000000    6256.000000     33.40000     17.000000      0.000000   
50%        0.000000   11868.500000     51.40000     24.000000      0.000000   
75%        0.000000   21223.250000     70.10000     32.000000      0.000000   
max       23.000000  647310.000000    134.30000    111.000000  22692.220000   

       out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
count   22510.000000  22510.000000     22510.000000     22510.000000   
mean      994.471550  15043.269607     15037.171987     12062.974932   
std      3017.074677   9788.105024      9782.869853      8384.981147   
min         0.000000      0.000000         0.000000         0.000000   
25%         0.000000   7359.465000      7357.160000      5670.825000   
50%         0.000000  12800.504927     12797.025000     10000.000000   
75%         0.000000  20863.242500     20860.462500     16500.000000   
max     22692.220000  57291.001148     57291.000000     35000.000000   

       total_rec_int  total_rec_late_fee    recoveries  \
count   225

Here are some of the key statistics:

- The average loan amount (loan_amnt) is approximately $15,031.8.
- The minimum loan amount is $1,000, and the maximum is $35,000.
- The average interest rate (int_rate) is approximately 12.255%.
- The average annual income (annual_inc) of borrowers is approximately $78,206.9.

In [14]:
# Check the number of null values in each column
df.isnull().sum()

id                                                0
member_id                                     22511
loan_amnt                                         0
funded_amnt                                       0
funded_amnt_inv                                   0
term                                              0
int_rate                                          0
installment                                       0
grade                                             0
sub_grade                                         0
emp_title                                      1349
emp_length                                     1340
home_ownership                                    0
annual_inc                                        0
verification_status                               0
issue_d                                           0
loan_status                                       0
pymnt_plan                                        0
url                                               0
desc        

In [15]:
# Check for duplicate rows
df.duplicated().sum()

0

## Cleaning

### Handling Missing Value

In [16]:
# Check the percentage of missing values in each column
df.isnull().sum() / len(df) * 100

id                                              0.000000
member_id                                     100.000000
loan_amnt                                       0.000000
funded_amnt                                     0.000000
funded_amnt_inv                                 0.000000
term                                            0.000000
int_rate                                        0.000000
installment                                     0.000000
grade                                           0.000000
sub_grade                                       0.000000
emp_title                                       5.992626
emp_length                                      5.952645
home_ownership                                  0.000000
annual_inc                                      0.000000
verification_status                             0.000000
issue_d                                         0.000000
loan_status                                     0.000000
pymnt_plan                     

In [17]:
# Create a copy of the original dataframe
df_clean = df.copy()

# Get the columns with more than 20% missing values
cols_to_drop = df_clean.columns[df_clean.isnull().sum() / len(df_clean) * 100 > 20]

# Drop the columns with more than 20% missing values
df_clean.drop(cols_to_drop, axis=1, inplace=True)

In [18]:
# Check the shape of the dataframe
df_clean.shape

(22511, 107)

In [19]:
# Check the percentage of missing values in each column
df_clean.isnull().sum() / len(df_clean) * 100

id                             0.000000
loan_amnt                      0.000000
funded_amnt                    0.000000
funded_amnt_inv                0.000000
term                           0.000000
int_rate                       0.000000
installment                    0.000000
grade                          0.000000
sub_grade                      0.000000
emp_title                      5.992626
emp_length                     5.952645
home_ownership                 0.000000
annual_inc                     0.000000
verification_status            0.000000
issue_d                        0.000000
loan_status                    0.000000
pymnt_plan                     0.000000
url                            0.000000
purpose                        0.000000
title                          0.590822
zip_code                       0.004442
addr_state                     0.004442
dti                            0.008885
delinq_2yrs                    0.004442
earliest_cr_line               0.004442


Let's investigate the columns with relatively high missing values comparing to the rest of the dataset:
1. il_util (~17% of the rows are missing)

2. mths_since_recent_inq (~10% of the rows are missing)

In [20]:
# Exam the il_util column
col_info('il_util')

Ratio of total current balance to high credit/credit limit on all install acct


**Interpretation of the column:**

The column "il_util" refers to a financial metric that calculates the percentage of the total current balance owed on all installment accounts in relation to the high credit or credit limit allowed on those accounts.

**Timing of Variable in the Loan Application Process:**

Arises **after** Loan Disbursement.

Information about the current balances on installment accounts is determined after the loan has been disbursed and the borrower starts using their installment accounts.

**Decision on data handling:**

Since it arises after loan disbursement, it is not useful for loan prediction analysis, let's **drop** this column.


In [21]:
# drop the il_util column
df_clean.drop('il_util', axis=1, inplace=True)

In [22]:
# Exam the mths_since_recent_inq column
col_info('mths_since_recent_inq')

Months since most recent inquiry.


**Interpretation of the column:**

The "mths_since_recent_inq" column in the LendingClub loan data tells us the time that has passed since the last hard credit inquiry was made on the borrower's credit report at the time of the loan application. 

It provides information about how actively the borrower has been seeking credit recently, and it can be a factor considered by LendingClub when evaluating a loan application. Generally, a longer period since the last inquiry is considered more favorable by lenders, as it suggests that the borrower is not taking on new credit obligations too frequently.

**Timing of Variable in the Loan Application Process:**

Arises **before** Loan Disbursement.

Credit inquiries occur when LendingClub check the borrower's credit report as part of the loan application process. It reflects the borrower's recent credit activity and credit-seeking behavior at the time of applying for the loan.

**Decision on data handling:**

Since it arises before loan disbursement, it is useful for loan prediction analysis, let's **keep** this column.

In [23]:
# Get the columns with less than 20% missing values
cols_to_fill = df_clean.columns[df_clean.isnull().sum() / len(df_clean) * 100 < 20]

# Fill the missing values with the mean of the respective columns for float columns
for col in cols_to_fill:
    if df_clean[col].dtype == 'float64':
        df_clean[col].fillna(df_clean[col].mean(), inplace=True)

# Check the percentage of missing values in each column 
df_clean.isnull().sum() / len(df_clean) * 100

id                            0.000000
loan_amnt                     0.000000
funded_amnt                   0.000000
funded_amnt_inv               0.000000
term                          0.000000
int_rate                      0.000000
installment                   0.000000
grade                         0.000000
sub_grade                     0.000000
emp_title                     5.992626
emp_length                    5.952645
home_ownership                0.000000
annual_inc                    0.000000
verification_status           0.000000
issue_d                       0.000000
loan_status                   0.000000
pymnt_plan                    0.000000
url                           0.000000
purpose                       0.000000
title                         0.590822
zip_code                      0.004442
addr_state                    0.004442
dti                           0.000000
delinq_2yrs                   0.000000
earliest_cr_line              0.004442
fico_range_low           

Let's investigate the columns with relatively high missing values comparing to the rest of the dataset:
1. emp_title (~6% of the rows are missing)

2. emp_length (~6% of the rows are missing)

In [24]:
# Exam the emp_title column
col_info('emp_title')

The job title supplied by the Borrower when applying for the loan.*


**Interpretation of the column:**

The "emp_title" column in the LendingClub loan data refers to the job title that the borrower provides on their loan application. When individuals apply for a loan with LendingClub, they are typically asked to provide various pieces of personal and financial information, including their employment details.

**Timing of Variable in the Loan Application Process:**

Arises **before** Loan Disbursement.

This information is used by the lender, LendingClub, as part of the loan application and underwriting process to assess the borrower's income stability, job status, and creditworthiness.

However, because the job title is self-reported, it is considered less reliable than verified income data, and lenders often rely on other financial factors to make their lending decisions.

**Decision on data handling:**

Although it arises before loan disbursement, but it is unrealistic for us to fill the missing job title for the borrower without any insider information. Therefore, let's **drop** the missing rows for this column.

In [25]:
# Exam the emp_length column
col_info('emp_length')

Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years. 


**Interpretation of the column:**

The "emp_length" column in the LendingClub loan data refers to the number of years the borrower has been employed at their current job or employer at the time of applying for the loan.

**Timing of Variable in the Loan Application Process:**

Arises **before** Loan Disbursement.

When individuals apply for a loan with LendingClub, they are typically required to provide information about their employment history, including the length of time they have been working at their current job.

However, the reported employment length is self-reported by the borrower and is not independently verified by the lender in most cases. It is just one of several factors considered during the loan application and underwriting process.

**Decision on data handling:**

Although it arises before loan disbursement, but it is unrealistic for us to fill the missing employment length for the borrower without any insider information. Therefore, let's **drop** the missing rows for this column.

In [26]:
# Drop the missing values for object columns
df_clean.dropna(inplace=True)

# Check the percentage of missing values in each column
df_clean.isnull().sum() / len(df_clean) * 100

id                            0.0
loan_amnt                     0.0
funded_amnt                   0.0
funded_amnt_inv               0.0
term                          0.0
int_rate                      0.0
installment                   0.0
grade                         0.0
sub_grade                     0.0
emp_title                     0.0
emp_length                    0.0
home_ownership                0.0
annual_inc                    0.0
verification_status           0.0
issue_d                       0.0
loan_status                   0.0
pymnt_plan                    0.0
url                           0.0
purpose                       0.0
title                         0.0
zip_code                      0.0
addr_state                    0.0
dti                           0.0
delinq_2yrs                   0.0
earliest_cr_line              0.0
fico_range_low                0.0
fico_range_high               0.0
inq_last_6mths                0.0
open_acc                      0.0
pub_rec       

In [27]:
# Check the shape of the dataframe after handling missing values
df_clean.shape

(21015, 106)

### Handling Features to Avoid Data Leakage

#### Post-loan Features

After carefulling examing the remaining columns in the dataset, these are the post-loan features that should be dropped since they arise after the loans are disbursed:

1.	total_pymnt
2.	total_pymnt_inv
3.	total_rec_prncp
4.	total_rec_int
5.	total_rec_late_fee
6.	recoveries
7.	collection_recovery_fee
8.	last_pymnt_d
9.	last_pymnt_amnt
10.	last_credit_pull_d
11.	last_fico_range_high
12.	last_fico_range_low
13.	collections_12_mths_ex_med
14.	application_type
15.	acc_now_delinq
16.	tot_coll_amt
17.	tot_cur_bal
18.	open_acc_6m
19.	open_act_il
20.	open_il_12m
21.	open_il_24m
22.	mths_since_rcnt_il
23.	total_bal_il
24.	open_rv_12m
25.	open_rv_24m
26.	max_bal_bc
27.	all_util
28.	total_rev_hi_lim
29.	inq_fi
30.	total_cu_tl
31.	inq_last_12m
32.	acc_open_past_24mths
33.	avg_cur_bal
34.	bc_open_to_buy
35.	bc_util
36.	chargeoff_within_12_mths
37.	delinq_amnt
38.	mo_sin_old_il_acct
39.	mo_sin_old_rev_tl_op
40.	mo_sin_rcnt_rev_tl_op
41.	mo_sin_rcnt_tl
42.	mort_acc
43.	mths_since_recent_bc
44.	mths_since_recent_inq
45.	num_accts_ever_120_pd
46.	num_actv_bc_tl
47.	num_actv_rev_tl
48.	num_bc_sats
49.	num_bc_tl
50.	num_il_tl
51.	num_op_rev_tl
52.	num_rev_accts
53.	num_rev_tl_bal_gt_0
54.	num_sats
55.	num_tl_120dpd_2m
56.	num_tl_30dpd
57.	num_tl_90g_dpd_24m
58.	num_tl_op_past_12m
59.	pct_tl_nvr_dlq
60.	percent_bc_gt_75
61.	pub_rec_bankruptcies
62.	tax_liens
63.	tot_hi_cred_lim
64.	total_bal_ex_mort
65.	total_bc_limit
66.	total_il_high_credit_limit
67.	hardship_flag
68.	disbursement_method
69.	debt_settlement_flag

These features are related to post-loan payment, recovery, credit history, and other metrics that are relevant after the loans have been disbursed.

Here is the description for each post-loan feature:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>LoanStatNew</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>total_pymnt</td>
      <td>Payments received to date for the total amount funded.</td>
    </tr>
    <tr>
      <th>2</th>
      <td>total_pymnt_inv</td>
      <td>Payments received to date for the portion of the total amount funded by investors.</td>
    </tr>
    <tr>
      <th>3</th>
      <td>total_rec_prncp</td>
      <td>Principal received to date.</td>
    </tr>
    <tr>
      <th>4</th>
      <td>total_rec_int</td>
      <td>Interest received to date.</td>
    </tr>
    <tr>
      <th>5</th>
      <td>total_rec_late_fee</td>
      <td>Late fees received to date.</td>
    </tr>
    <tr>
      <th>6</th>
      <td>recoveries</td>
      <td>Post charge-off gross recovery.</td>
    </tr>
    <tr>
      <th>7</th>
      <td>collection_recovery_fee</td>
      <td>Post charge-off collection fee.</td>
    </tr>
    <tr>
      <th>8</th>
      <td>last_pymnt_d</td>
      <td>Last month payment was received.</td>
    </tr>
    <tr>
      <th>9</th>
      <td>last_pymnt_amnt</td>
      <td>Last total payment amount received.</td>
    </tr>
    <tr>
      <th>10</th>
      <td>last_credit_pull_d</td>
      <td>The most recent month LC pulled credit for this loan.</td>
    </tr>
    <tr>
      <th>11</th>
      <td>last_fico_range_high</td>
      <td>The upper boundary range the borrower's last FICO pulled belongs to.</td>
    </tr>
    <tr>
      <th>12</th>
      <td>last_fico_range_low</td>
      <td>The lower boundary range the borrower's last FICO pulled belongs to.</td>
    </tr>
    <tr>
      <th>13</th>
      <td>collections_12_mths_ex_med</td>
      <td>Number of collections in 12 months excluding medical collections.</td>
    </tr>
    <tr>
      <th>14</th>
      <td>application_type</td>
      <td>Indicates whether the loan is an individual application or a joint application with two co-borrowers.</td>
    </tr>
    <tr>
      <th>15</th>
      <td>acc_now_delinq</td>
      <td>The number of accounts on which the borrower is now delinquent.</td>
    </tr>
    <tr>
      <th>16</th>
      <td>tot_coll_amt</td>
      <td>Total collection amounts ever owed.</td>
    </tr>
    <tr>
      <th>17</th>
      <td>tot_cur_bal</td>
      <td>Total current balance of all accounts.</td>
    </tr>
    <tr>
      <th>18</th>
      <td>open_acc_6m</td>
      <td>Number of open trades in the last 6 months.</td>
    </tr>
    <tr>
      <th>19</th>
      <td>open_act_il</td>
      <td>Number of currently active installment trades.</td>
    </tr>
    <tr>
      <th>20</th>
      <td>open_il_12m</td>
      <td>Number of installment accounts opened in the past 12 months.</td>
    </tr>
    <tr>
      <th>21</th>
      <td>open_il_24m</td>
      <td>Number of installment accounts opened in the past 24 months.</td>
    </tr>
    <tr>
      <th>22</th>
      <td>mths_since_rcnt_il</td>
      <td>Months since most recent installment accounts opened.</td>
    </tr>
    <tr>
      <th>23</th>
      <td>total_bal_il</td>
      <td>Total current balance of all installment accounts.</td>
    </tr>
    <tr>
      <th>24</th>
      <td>open_rv_12m</td>
      <td>Number of revolving trades opened in the past 12 months.</td>
    </tr>
    <tr>
      <th>25</th>
      <td>open_rv_24m</td>
      <td>Number of revolving trades opened in the past 24 months.</td>
    </tr>
    <tr>
      <th>26</th>
      <td>max_bal_bc</td>
      <td>Maximum current balance owed on all revolving accounts.</td>
    </tr>
    <tr>
      <th>27</th>
      <td>all_util</td>
      <td>Balance to credit limit on all trades.</td>
    </tr>
    <tr>
      <th>28</th>
      <td>total_rev_hi_lim</td>
      <td>Total bank card high credit/credit limit.</td>
    </tr>
    <tr>
      <th>29</th>
      <td>inq_fi</td>
      <td>Number of personal finance inquiries.</td>
    </tr>
    <tr>
      <th>30</th>
      <td>total_cu_tl</td>
      <td>Number of finance trades.</td>
    </tr>
    <tr>
      <th>31</th>
      <td>inq_last_12m</td>
      <td>Number of credit inquiries in the past 12 months.</td>
    </tr>
    <tr>
      <th>32</th>
      <td>acc_open_past_24mths</td>
      <td>Number of trades opened in the past 24 months.</td>
    </tr>
    <tr>
      <th>33</th>
      <td>avg_cur_bal</td>
      <td>Average current balance of all accounts.</td>
    </tr>
    <tr>
      <th>34</th>
      <td>bc_open_to_buy</td>
      <td>Total open to buy on revolving bank cards.</td>
    </tr>
    <tr>
      <th>35</th>
      <td>bc_util</td>
      <td>Ratio of total current balance to high credit/credit limit for all bank card accounts.</td>
    </tr>
    <tr>
      <th>36</th>
      <td>chargeoff_within_12_mths</td>
      <td>Number of charge-offs within 12 months.</td>
    </tr>
    <tr>
      <th>37</th>
      <td>delinq_amnt</td>
      <td>The past-due amount owed for the accounts on which the borrower is now delinquent.</td>
    </tr>
    <tr>
      <th>38</th>
      <td>mo_sin_old_il_acct</td>
      <td>Months since oldest bank installment account opened.</td>
    </tr>
    <tr>
      <th>39</th>
      <td>mo_sin_old_rev_tl_op</td>
      <td>Months since oldest revolving account opened.</td>
    </tr>
    <tr>
      <th>40</th>
      <td>mo_sin_rcnt_rev_tl_op</td>
      <td>Months since most recent revolving account opened.</td>
    </tr>
    <tr>
      <th>41</th>
      <td>mo_sin_rcnt_tl</td>
      <td>Months since most recent account opened.</td>
    </tr>
    <tr>
      <th>42</th>
      <td>mort_acc</td>
      <td>Number of mortgage accounts.</td>
    </tr>
    <tr>
      <th>43</th>
      <td>mths_since_recent_bc</td>
      <td>Months since most recent bank card account opened.</td>
    </tr>
    <tr>
      <th>44</th>
      <td>mths_since_recent_inq</td>
      <td>Months since most recent inquiry.</td>
    </tr>
    <tr>
      <th>45</th>
      <td>num_accts_ever_120_pd</td>
      <td>Number of accounts ever 120 or more days past due.</td>
    </tr>
    <tr>
      <th>46</th>
      <td>num_actv_bc_tl</td>
      <td>Number of currently active bank card accounts.</td>
    </tr>
    <tr>
      <th>47</th>
      <td>num_actv_rev_tl</td>
      <td>Number of currently active revolving trades.</td>
    </tr>
    <tr>
      <th>48</th>
      <td>num_bc_sats</td>
      <td>Number of satisfactory bank card accounts.</td>
    </tr>
    <tr>
      <th>49</th>
      <td>num_bc_tl</td>
      <td>Number of bank card accounts.</td>
    </tr>
    <tr>
      <th>50</th>
      <td>num_il_tl</td>
      <td>Number of installment accounts.</td>
    </tr>
    <tr>
      <th>51</th>
      <td>num_op_rev_tl</td>
      <td>Number of open revolving accounts.</td>
    </tr>
    <tr>
      <th>52</th>
      <td>num_rev_accts</td>
      <td>Number of revolving accounts.</td>
    </tr>
    <tr>
      <th>53</th>
      <td>num_rev_tl_bal_gt_0</td>
      <td>Number of revolving trades with a balance >0.</td>
    </tr>
    <tr>
      <th>54</th>
      <td>num_sats</td>
      <td>Number of satisfactory accounts.</td>
    </tr>
    <tr>
      <th>55</th>
      <td>num_tl_120dpd_2m</td>
      <td>Number of accounts currently 120 days past due (updated in the past 2 months).</td>
    </tr>
    <tr>
      <th>56</th>
      <td>num_tl_30dpd</td>
      <td>Number of accounts currently 30 days past due (updated in the past 2 months).</td>
    </tr>
    <tr>
      <th>57</th>
      <td>num_tl_90g_dpd_24m</td>
      <td>Number of accounts 90 or more days past due in the last 24 months.</td>
    </tr>
    <tr>
      <th>58</th>
      <td>num_tl_op_past_12m</td>
      <td>Number of accounts opened in the past 12 months.</td>
    </tr>
    <tr>
      <th>59</th>
      <td>pct_tl_nvr_dlq</td>
      <td>Percent of trades never delinquent.</td>
    </tr>
    <tr>
      <th>60</th>
      <td>percent_bc_gt_75</td>
      <td>Percentage of all bank card accounts > 75% of the limit.</td>
    </tr>
    <tr>
      <th>61</th>
      <td>pub_rec_bankruptcies</td>
      <td>Number of public record bankruptcies.</td>
    </tr>
    <tr>
      <th>62</th>
      <td>tax_liens</td>
      <td>Number of tax liens.</td>
    </tr>
    <tr>
      <th>63</th>
      <td>tot_hi_cred_lim</td>
      <td>Total high credit/credit limit.</td>
    </tr>
    <tr>
      <th>64</th>
      <td>total_bal_ex_mort</td>
      <td>Total credit balance excluding mortgage.</td>
    </tr>
    <tr>
      <th>65</th>
      <td>total_bc_limit</td>
      <td>Total bank card high credit/credit limit.</td>
    </tr>
    <tr>
      <th>66</th>
      <td>total_il_high_credit_limit</td>
      <td>Total installment high credit/credit limit.</td>
    </tr>
    <tr>
      <th>67</th>
      <td>hardship_flag</td>
      <td>Flags whether or not the borrower is on a hardship plan.</td>
    </tr>
    <tr>
      <th>68</th>
      <td>disbursement_method</td>
      <td>The method by which the borrower receives their loan.</td>
    </tr>
    <tr>
      <th>69</th>
      <td>debt_settlement_flag</td>
      <td>Flags whether or not the borrower has settled their debt.</td>
    </tr>
  </tbody>
</table>

In [28]:
# List of columns to drop (post-loan features)
columns_to_drop = [
    'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
    'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d',
    'last_pymnt_amnt', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low',
    'collections_12_mths_ex_med', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
    'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m',
    'mths_since_rcnt_il', 'total_bal_il', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
    'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
    'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths',
    'delinq_amnt', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op',
    'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq',
    'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats',
    'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 'num_rev_accts', 'num_rev_tl_bal_gt_0',
    'num_sats', 'num_tl_120dpd_2m', 'num_tl_30dpd', 'num_tl_90g_dpd_24m',
    'num_tl_op_past_12m', 'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pub_rec_bankruptcies',
    'tax_liens', 'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit',
    'total_il_high_credit_limit', 'hardship_flag', 'disbursement_method', 'debt_settlement_flag'
]

# Drop the columns from the DataFrame
df_clean.drop(columns=columns_to_drop, inplace=True)

In [29]:
# Check the shape of the dataframe after dropping post-loan features
df_clean.shape

(21015, 37)

In [30]:
# List the remaining columns
df_clean.columns.tolist()

['id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'policy_code']

#### Irrelevant/Redundant Features

There are some features that can be dropped since they do not contribute to the analysis or are considered as redundant.

To determine which features do not contribute to the analysis or are redundant for predicting the loan_status, we need to consider the relevance and potential impact of each feature on the target variable. Based on the information provided, the following features seem less likely to directly contribute to the analysis:

1. **id:** This variable appears to be an identifier for the loan and does not provide any meaningful information for predicting loan_status.

2. **url:** This variable is likely a link to the loan's webpage and is not relevant to predicting loan_status.

3. **pymnt_plan:** Since most of the data points may have the same value (e.g., "n" for no), this variable may not have enough variability to be useful in the analysis.

4. **title:** The loan title might contain similar information as the 'purpose' variable and may not add much additional value for prediction.

5. **zip_code:** The geographic location of borrowers might not have a strong correlation with loan_status.

6. **initial_list_status:** This variable might not have a significant impact on loan_status prediction.

7. **out_prncp** and **out_prncp_inv:** These variables represent outstanding principal amount and are likely to have a strong correlation with the loan_status, making them less useful for prediction.

8. **policy_code:** If most of the data points have the same value (e.g., "1"), then this variable may not provide any meaningful information for the analysis.

Here is the description for the above features:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>LoanStatNew</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>id</td>
      <td>A unique LC assigned ID for the loan listing.</td>
    </tr>
    <tr>
      <th>2</th>
      <td>url</td>
      <td>URL for the LC page with listing data.</td>
    </tr>
    <tr>
      <th>3</th>
      <td>pymnt_plan</td>
      <td>Indicates if a payment plan has been put in place for the loan.</td>
    </tr>
    <tr>
      <th>4</th>
      <td>title</td>
      <td>The loan title provided by the borrower.</td>
    </tr>
    <tr>
      <th>5</th>
      <td>zip_code</td>
      <td>The first 3 numbers of the zip code provided by the borrower in the loan application.</td>
    </tr>
    <tr>
      <th>6</th>
      <td>initial_list_status</td>
      <td>The initial listing status of the loan. Possible values are "w" (whole) or "f" (fraction).</td>
    </tr>
    <tr>
      <th>7</th>
      <td>out_prncp</td>
      <td>Remaining outstanding principal amount for total amount funded.</td>
    </tr>
    <tr>
      <th>8</th>
      <td>out_prncp_inv</td>
      <td>Remaining outstanding principal amount for portion of total amount funded by investors.</td>
    </tr>
    <tr>
      <th>9</th>
      <td>policy_code</td>
      <td>Publicly available policy_code=1, new products not publicly available policy_code=2.</td>
    </tr>
  </tbody>
</table>

Let's further investigate some of the features mentioned above:

In [31]:
# Check unique values in each column
variables_list = ['pymnt_plan', 'title', 'purpose', 'initial_list_status', 'policy_code']

# Initialize a dictionary to store the counts for each variable
variable_counts = {}

# Loop through each variable in the variables_list
for var in variables_list:
    # Use value_counts() to count the occurrences of each unique value and store it in the dictionary
    variable_counts[var] = df[var].value_counts()

# Print the counts for each variable
for var, counts in variable_counts.items():
    print(f"Counts for '{var}':")
    print(counts)
    print()

Counts for 'pymnt_plan':
n    22508
y        3
Name: pymnt_plan, dtype: int64

Counts for 'title':
Debt consolidation         12676
Credit card refinancing     5621
Home improvement            1269
Other                       1242
Major purchase               478
Medical expenses             255
Business                     236
Car financing                205
Vacation                     146
Moving and relocation        143
Home buying                   93
Green loan                    14
Name: title, dtype: int64

Counts for 'purpose':
debt_consolidation    12742
credit_card            5652
home_improvement       1284
other                  1253
major_purchase          480
medical                 257
small_business          238
car                     208
vacation                145
moving                  144
house                    93
renewable_energy         14
debt_cons                 1
Name: purpose, dtype: int64

Counts for 'initial_list_status':
w    19567
f     2943
Name: i

Here we have confirmed with further information from the dataset that the following features are less likely to directly contribute to the analysis, so we are going to drop them:

1. id
2. url
3. pymnt_plan
4. title
5. zip_code
6. initial_list_status
7. out_prncp & out_prncp_inv
8. policy_code

In [32]:
# List of columns to drop
cols_to_drop = ['id', 'url', 'pymnt_plan', 'title', 'zip_code', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'policy_code']

# Use the drop() method to remove the specified columns
df_clean = df_clean.drop(columns=cols_to_drop)

In [33]:
# Check the shape of the dataframe after dropping post-loan features
df_clean.shape

(21015, 28)

In [34]:
# List the remaining columns
df_clean.columns.tolist()

['loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'purpose',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc']

In [35]:
# Show a sample of the data
df_clean.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0     3600.0       3600.0           3600.0   36 months     13.99       123.03   
1    24700.0      24700.0          24700.0   36 months     11.99       820.28   
3    35000.0      35000.0          35000.0   60 months     14.85       829.90   
4    10400.0      10400.0          10400.0   60 months     22.45       289.91   
5    11950.0      11950.0          11950.0   36 months     13.44       405.18   

  grade sub_grade                    emp_title emp_length home_ownership  \
0     C        C4                      leadman  10+ years       MORTGAGE   
1     C        C1                     Engineer  10+ years       MORTGAGE   
3     C        C5  Information Systems Officer  10+ years       MORTGAGE   
4     F        F1          Contract Specialist    3 years       MORTGAGE   
5     C        C3         Veterinary Tecnician    4 years           RENT   

   annual_inc verification_status   issue_d loan_status             purpose  \
0     55000.0        Not Verified  Dec-2015  Fully Paid  debt_consolidation   
1     65000.0        Not Verified  Dec-2015  Fully Paid      small_business   
3    110000.0     Source Verified  Dec-2015     Current  debt_consolidation   
4    104433.0     Source Verified  Dec-2015  Fully Paid      major_purchase   
5     34000.0     Source Verified  Dec-2015  Fully Paid  debt_consolidation   

  addr_state    dti  delinq_2yrs earliest_cr_line  fico_range_low  \
0         PA   5.91          0.0         Aug-2003           675.0   
1         SD  16.06          1.0         Dec-1999           715.0   
3         NJ  17.06          0.0         Sep-2008           785.0   
4         PA  25.37          1.0         Jun-1998           695.0   
5         GA  10.20          0.0         Oct-1987           690.0   

   fico_range_high  inq_last_6mths  open_acc  pub_rec  revol_bal  revol_util  \
0            679.0             1.0       7.0      0.0     2765.0        29.7   
1            719.0             4.0      22.0      0.0    21470.0        19.2   
3            789.0             0.0      13.0      0.0     7802.0        11.6   
4            699.0             3.0      12.0      0.0    21929.0        64.5   
5            694.0             0.0       5.0      0.0     8822.0        68.4   

   total_acc  
0       13.0  
1       38.0  
3       17.0  
4       35.0  
5        6.0

In [36]:
# Save the cleaned data to a parquet file
df_clean.to_parquet('dataset/LoanData_after_DataCleaning.parquet')